Allow in-place modifications of representations

mhvk opened the pull request:
This is a follow-up of #9857 ~(EDIT: and builds upon it, hence the many commits)~, making also representations mutable. As commented there, representations themselves have no reason not to be mutable, since the behaviour of setting is well defined (in a way, they are just glorified quantities). Now that coordinates have become mutable, it makes little sense to keep the representations immutable.

~I hope to follow this up with an `DataInfo` class such that representations can also be added to tables; without this, they would be second-class citizens.~ EDIT: now added

fixes #6447 (as a side effect)

mhvk commented:
@taldcroft - I think this is ready! Could you review? Especially the stuff related to table integration (i.e., `info`, small changes in `operations.py` and `yaml.py`).

taldcroft commented on astropy/table/tests/conftest.py:
```diff
@@ -137,13 +137,22 @@ def table_type(request):
               'time': time.Time([2000, 2001, 2002, 2003], format='jyear'),
               'skycoord': coordinates.SkyCoord(ra=[0, 1, 2, 3] * u.deg,
                                                dec=[0, 1, 2, 3] * u.deg),
+              'sphericalrep': coordinates.SphericalRepresentation(
```
Is this enough to test that roundtrip through FITS or ECSV works?  I forget...

mhvk replied:
I fear you are right that it is not...

mhvk replied:
now fixed

taldcroft commented on astropy/io/misc/yaml.py:
```diff
@@ -299,6 +299,16 @@ def ignore_aliases(self, data):
     AstropyLoader.add_constructor(tag, _quantity_constructor(cls))
 
 
+for cls in (list(coords.representation.REPRESENTATION_CLASSES.values())
```
Add an explicit test of YAML load(dump(obj)) roundtrip.

mhvk replied:
Added yaml tests. And, yes, it fixes #6447!

mhvk commented on CHANGES.rst:
```diff
@@ -22,6 +22,9 @@ astropy.coordinates
   coordinate object of the same class and equivalent frame attributes. [#9857]
 
+- Allow in-place modification of array-valued ``Representation`` objects,
+  including of possible included differentials. [#10210]
```
Done.

mhvk commented on CHANGES.rst:
```diff
@@ -113,6 +116,8 @@ astropy.table
 - Added support for ``SkyCoord`` mixin columns in ``dstack``, ``vstack`` and
   ``insert_row`` functions. [#9857]
 
+- Allow support for ``Representation`` and ``Differential`` mixin columns. [#10210]
```
Done.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -122,28 +196,52 @@ class BaseRepresentationOrDifferential(ShapedLikeNDArray):
-            except KeyError:
-                raise TypeError('__init__() missing 1 required positional '
-                                'argument: {!r}'.format(component))
+        if (args and isinstance(args[0], self.__class__)
```
Good point. I put it under API change.

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -44,6 +45,30 @@ def teardown_function(func):
+    return result
+
+
+def representation_equal(rep1, rep2):
```
Yes, it should be, but I thought to leave that for another PR (#10154??), in particular as I am not quite sure yet what to do with class mismatches.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -105,6 +106,79 @@ def _combine_xyz(x, y, z, xyz_axis=0):
+    _supports_indexing = False
+
+    @staticmethod
+    def default_format(val):
```
Indeed, done.

taldcroft commented:
None of this should be considered a blocker for this PR, but there are outstanding issues related to serialization that should either be fixed (as long as you are in the code) or transferred to separate issues:

The differentials are still not being serialized at all in a SkyCoord:
```
>>> yaml.load(yaml.dump(sc))  # Loses the differential
<SkyCoord (ICRS): (ra, dec, distance) in (deg, deg, Mpc)
    [(270., 30., 10.)]>

>>> sc = SkyCoord([270] * u.deg, [30] * u.deg, [10] * u.Mpc, radial_velocity=[20] * u.km / u.s)
>>> t = Table([sc])
>>> t.write('junk.ecsv', overwrite=True)
>>> t2 = Table.read('junk.ecsv')
>>> t2['col0']
<SkyCoord (ICRS): (ra, dec, distance) in (deg, deg, Mpc)
    [(270., 30., 10.)]>
>>> t['col0']
<SkyCoord (ICRS): (ra, dec, distance) in (deg, deg, Mpc)
    [(270., 30., 10.)]
 (radial_velocity) in km / s
    [(20.,)]>
```

For a representation in a table the differential is being serialized as an encoded binary numpy instead of getting dropped in as a SerializedColumn.  From a quick look we would need a deeper structure for the column name like `col0.differentials.s.d_x`.  What is confusing is that the x differential is called `v_x` on input but `d_x` in the differential.  No idea what's happening there.
```
>>> sc = SkyCoord(x=[1]*u.pc, y=[2]*u.pc, z=[3]*u.pc,
... ...      v_x=[4]*u.km/u.s, v_y=[5]*u.km/u.s, v_z=[6]*u.km/u.s,
... ...      representation_type=CartesianRepresentation,
... ...      differential_type=CartesianDifferential)
...
>>> t = Table([sc.frame.data])
>>> t.write('junk_repr.ecsv', overwrite=True)
>>> cat junk_repr.ecsv
# %ECSV 0.9
# ---
# datatype:
# - {name: col0.x, unit: pc, datatype: float64}
# - {name: col0.y, unit: pc, datatype: float64}
# - {name: col0.z, unit: pc, datatype: float64}
# meta: !!omap
# - __serialized_columns__:
#     col0:
#       __class__: astropy.coordinates.representation.CartesianRepresentation
#       __info__: {unit: pc}
#       differentials:
#         s: !astropy.coordinates.CartesianDifferential
#           d_x: !astropy.units.Quantity
#             unit: !astropy.units.Unit {unit: km / s}
#             value: !numpy.ndarray
#               buffer: !!binary |
#                 QUFBQUFBQUFFRUE9
#               dtype: float64
#               order: C
#               shape: !!python/tuple [1]
#           d_y: !astropy.units.Quantity
#             unit: !astropy.units.Unit {unit: km / s}
#             value: !numpy.ndarray
#               buffer: !!binary |
#                 QUFBQUFBQUFGRUE9
#               dtype: float64
#               order: C
#               shape: !!python/tuple [1]
#           d_z: !astropy.units.Quantity
#             unit: !astropy.units.Unit {unit: km / s}
#             value: !numpy.ndarray
#               buffer: !!binary |
#                 QUFBQUFBQUFHRUE9
#               dtype: float64
#               order: C
#               shape: !!python/tuple [1]
#       x: !astropy.table.SerializedColumn
#         __class__: astropy.units.quantity.Quantity
#         unit: &id001 !astropy.units.Unit {unit: pc}
#         value: !astropy.table.SerializedColumn {name: col0.x}
#       y: !astropy.table.SerializedColumn
#         __class__: astropy.units.quantity.Quantity
#         unit: *id001
#         value: !astropy.table.SerializedColumn {name: col0.y}
#       z: !astropy.table.SerializedColumn
#         __class__: astropy.units.quantity.Quantity
#         unit: *id001
#         value: !astropy.table.SerializedColumn {name: col0.z}
# schema: astropy-2.0
col0.x col0.y col0.z
1.0 2.0 3.0
```

mhvk commented:
Thanks, I had started to notice that something was odd with differentials, but hadn't quite understood what was happening yet, so this is very helpful. 

p.s. There also seems to be a strange problem with partial name clashes in serialized columns. If I can confirm that, I'll make it a separate issue, since it would need back-porting.

mhvk commented:
@taldcroft - OK, with your fix, I now could get the round-tripping in FITS/HDF5/ECSV to work. I think I also addressed all your comments, except that I couldn't easily get `SkyCoord` with differentials to work (or rather, I got half-way, but found that even when there is just a `radial_velocity`, magic `pm_ra_cosdec` and `pm_dec` appear; since this PR is big as is, I thought I better leave it as a separate issue...).

mhvk commented:
The failures are due to a strange but clearly temporary ability to fetch `extension_helpers`. I'll have another look at my own work here, and after that will restart CI just to be sure, and then merge.

Thanks, @adrn and @taldcroft for reviews. @eteq - time to speak up if you disagree with representations becoming settable is now...

taldcroft commented:
I'm seeing the same weird CI problems.

pllim commented:
Re: CI -- downgrade doctestplus to 0.5 + pytest 5.3

mhvk commented:
@pllim - ah, good to know. Is it now fixed (is it you who restarted circle-ci? I don't seem to have permission myself)

pllim commented:
I didn't touch it. 😄

bsipocz commented:
> ah, good to know. Is it now fixed (is it you who restarted circle-ci? I don't seem to have permission myself)

I think everyone with write access should have the power for circleCI. But you need to log-in with github to circleCI, and maybe also watch the repo. 

There should not be any failure any more as part of an upstream hackery I reverted the doctestplus release that was causing the trouble.

mhvk commented:
Darn, the windows run consistently hangs in `test_io_fast_c_reader`. Anybody any ideas?

pllim commented:
> windows run consistently hangs

Try run it with `pytest-faulthandler`.

mhvk commented:
@pllim - that sounds good, but how do I do that?

pllim commented:
@mhvk , I pushed a commit directly to the branch for the `pytest-faulthandler` and cancelled CircleCI but I don't know why Travis CI did not start. 😞

mhvk commented:
@pllim - interesting, thanks! But weird, what happened to the travis run!?

pllim commented:
You might need to rebase for Travis to run -- https://docs.travis-ci.com/user/pull-requests/#my-pull-request-isnt-being-built

mhvk commented:
OK, I saw a tiny mistake anyway, so I'll rebase keeping your commit at the top.

mhvk commented:
OK, travis is now running (and logging out and back in again to circle-CI meant I could indeed cancel).

mhvk commented:
And now I have to wait for the travis time-out... I hope this isn't going to take a number of iterations!

mhvk commented:
Hmm, that didn't help: still get
```
..\..\.tox\py37-test-alldeps\lib\site-packages\astropy\io\ascii\tests\test_c_reader.py x [ 17%]

..x.x.x.x.x.x.x.x.x.x.x.x...x................x.x.x.x.x.x.x.

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

The build has been terminated
```
From trying to count tests, I think it  is https://github.com/astropy/astropy/blob/986d5b74145c4cf6ee47c303d709832d199df24f/astropy/io/ascii/tests/test_c_reader.py#L660
so I'll go ahead and just try skipping this.

This is painful! I hate windows and hard-to-debug stuff due to compiled code!

mhvk commented:
OK, I'm getting too annoyed and tired for my own good, so leaving this to tomorrow. It was approved already so hopefully this can be treated as "bug fix" (wish I even had a clue why the fast c reader cared about anything I did...)

pllim commented:
Hmm, the `pytest-faulthandler` didn't smoke out anything; it timed out. Why does this feel familiar. When I get the chance, maybe I can try this PR on my Windows locally.

pllim commented:
So... It hangs on my Windows machine too. But when I run only `pytest astropy/io/ascii/tests/test_c_reader.py`, I get this (the weird slashes due to unholy union of running Git Bash on Windows):

```
=================================== ERRORS ====================================
__________ ERROR collecting astropy/io/ascii/tests/test_c_reader.py ___________
astropy\time\core.py:2590: in update_leap_seconds
    from astropy.utils import iers
astropy\utils\iers\__init__.py:2: in <module>
    from .iers import *
astropy\utils\iers\iers.py:26: in <module>
    from astropy.io.ascii import convert_numpy
E   ImportError: cannot import name 'convert_numpy' from 'astropy.io.ascii' (d:\...\astropy\io\ascii\__init__.py)

During handling of the above exception, another exception occurred:
c:\...\pytest_doctestplus\plugin.py:173: in collect
    module = self.fspath.pyimport()
c:\...\py\_path\local.py:668: in pyimport
    __import__(modname)
astropy\io\ascii\__init__.py:7: in <module>
    from .core import (InconsistentTableError,
astropy\io\ascii\core.py:30: in <module>
    from astropy.table import Table
astropy\table\__init__.py:54: in <module>
    from .serialize import SerializedColumn, represent_mixins_as_columns
astropy\table\serialize.py:10: in <module>
    from astropy.coordinates import representation as coorep
astropy\coordinates\__init__.py:16: in <module>
    from .builtin_frames import *
astropy\coordinates\builtin_frames\__init__.py:29: in <module>
    from .fk5 import FK5
astropy\coordinates\builtin_frames\fk5.py:8: in <module>
    from astropy.coordinates import earth_orientation as earth
astropy\coordinates\earth_orientation.py:19: in <module>
    jd1950 = Time('B1950').jd
astropy\time\core.py:395: in __new__
    update_leap_seconds()
astropy\time\core.py:2597: in update_leap_seconds
    f"exception: {exc!r}", AstropyWarning)
E   astropy.utils.exceptions.AstropyWarning: leap-second auto-update failed due to the following exception: ImportError("cannot import name 'convert_numpy' from 'astropy.io.ascii' (d:\\...\\astropy\\io\\ascii\\__init__.py)")
__________ ERROR collecting astropy/io/ascii/tests/test_c_reader.py ___________
ImportError while importing test module 'D:\...\astropy\io\ascii\tests\test_c_reader.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
astropy\io\ascii\__init__.py:7: in <module>
    from .core import (InconsistentTableError,
astropy\io\ascii\core.py:30: in <module>
    from astropy.table import Table
astropy\table\__init__.py:54: in <module>
    from .serialize import SerializedColumn, represent_mixins_as_columns
astropy\table\serialize.py:10: in <module>
    from astropy.coordinates import representation as coorep
astropy\coordinates\__init__.py:16: in <module>
    from .builtin_frames import *
astropy\coordinates\builtin_frames\__init__.py:29: in <module>
    from .fk5 import FK5
astropy\coordinates\builtin_frames\fk5.py:11: in <module>
    from .utils import EQUINOX_J2000
astropy\coordinates\builtin_frames\utils.py:15: in <module>
    from astropy.utils import iers
astropy\utils\iers\__init__.py:2: in <module>
    from .iers import *
astropy\utils\iers\iers.py:26: in <module>
    from astropy.io.ascii import convert_numpy
E   ImportError: cannot import name 'convert_numpy' from 'astropy.io.ascii' (d:\...\astropy\io\ascii\__init__.py)
!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!
============================== 2 errors in 3.22s ==============================
```

With this PR, in a Python session I am unable to do `from astropy.io.ascii import convert_numpy` on Windows but it works with latest `master`. 🤷

pllim commented:
Pretty sure this is a case of circular import offending Windows. I'll push a potential fix out.

pllim commented:
p.s. I don't think remote data failures are related but you should double check.

mhvk commented:
Wow, @pllim, that was some dedicated sleuthing! Thanks, so much!

As you note, the remote-date failures are all unrelated, so I'll go ahead and merge.  It must still be friday somewhere...

taldcroft commented:
@mhvk - this PR is breaking my local astropy.  I get this from a clean build of astropy with this PR merged, and do not see a problem beforehand.  This is with numpy 1.18.1.
```
In [1]: astro                                                                                                                                                 
astropy=4.1.dev1333+g27983b59c
/Users/aldcroft/git/astropy/astropy/table/column.py:1020: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = getattr(super(), op)(other)
Downloading https://hpiers.obspm.fr/iers/bul/bulc/Leap_Second.dat
|========================================================================================================================| 1.3k/1.3k (100.00%)         0s
Downloading https://www.ietf.org/timezones/data/leap-seconds.list
|========================================================================================================================|  10k/ 10k (100.00%)         0s
WARNING: IERSStaleWarning: leap-second file is expired. [astropy.utils.iers.iers]
```

Slightly OT, but the whole leap second download stuff still seems to be a source of really inscrutable problems.  I really think it would be worth something like a config option to say "Don't ever try leap second downloads".  I use astropy on mission-critical and non-networked computers, and I don't want it failing on that account.

pllim commented:
Re: FutureWarning -- Hmm, why didn't the CI catch it?

taldcroft commented:
Another manifestation:
```
(astropy) ➜  astropy git:(5ff936bd2) pytest astropy/io/ascii
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.6.10, pytest-5.3.5, py-1.8.1, pluggy-0.13.1

Running tests with Astropy version 4.1.dev1494+g5ff936bd2.
Running tests in astropy/io/ascii.

Date: 2020-05-02T11:53:44

Platform: Darwin-19.2.0-x86_64-i386-64bit

Executable: /Users/aldcroft/miniconda3/envs/astropy/bin/python

Full Python Version: 
3.6.10 |Anaconda, Inc.| (default, Jan  7 2020, 15:01:53) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.18.1
Scipy: 1.4.1
Matplotlib: 3.1.3
h5py: 2.10.0
Pandas: 1.0.0
Cython: not available
Scikit-image: 0.16.2
asdf: 2.6.0

Using Astropy options: remote_data: none.

rootdir: /Users/aldcroft/git/astropy, inifile: setup.cfg
plugins: arraydiff-0.3, remotedata-0.3.2, hypothesis-5.4.1, openfiles-0.4.0, xdist-1.31.0, filter-subpackage-0.1.1, forked-1.1.3, doctestplus-0.5.0, astropy-header-0.1.2, cov-2.8.1, asdf-2.6.0
collected 576 items / 1 error / 575 selected                                                                                                                 

=========================================================================== ERRORS ===========================================================================
_______________________________________________________ ERROR collecting astropy/io/ascii/__init__.py ________________________________________________________
astropy/time/core.py:2592: in update_leap_seconds
    table = iers.LeapSeconds.auto_open(files)
astropy/utils/iers/iers.py:1027: in auto_open
    warn('leap-second file is expired.', IERSStaleWarning)
E   astropy.utils.iers.iers.IERSStaleWarning: leap-second file is expired.

During handling of the above exception, another exception occurred:
astropy/io/ascii/__init__.py:7: in <module>
    from .core import (InconsistentTableError,
astropy/io/ascii/core.py:30: in <module>
    from astropy.table import Table
astropy/table/__init__.py:54: in <module>
    from .serialize import SerializedColumn, represent_mixins_as_columns
astropy/table/serialize.py:10: in <module>
    from astropy.coordinates import representation as coorep
astropy/coordinates/__init__.py:16: in <module>
    from .builtin_frames import *
astropy/coordinates/builtin_frames/__init__.py:29: in <module>
    from .fk5 import FK5
astropy/coordinates/builtin_frames/fk5.py:8: in <module>
    from astropy.coordinates import earth_orientation as earth
astropy/coordinates/earth_orientation.py:19: in <module>
    jd1950 = Time('B1950').jd
astropy/time/core.py:395: in __new__
    update_leap_seconds()
astropy/time/core.py:2597: in update_leap_seconds
    f"exception: {exc!r}", AstropyWarning)
E   astropy.utils.exceptions.AstropyWarning: leap-second auto-update failed due to the following exception: IERSStaleWarning('leap-second file is expired.',)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
====================================================================== 1 error in 1.86s ======================================================================
```

taldcroft commented:
So the question is if anyone else can reproduce this or if it's just me.

mhvk commented:
I don't see how this could be related to this PR - but agree with the sentiment that the `LeapSecond` stuff is too fragile -- #9479.  I'd very much appreciate help in trying to get it better!! While writing it, I found myself in an awful circular import morass between `iers` needing `Time`, and `Time` needing `iers`. I think for tests we should silence the stale-warning, since it is now turned into an error (which didn't use to be the case when I wrote the stuff).

It is possible to avoid any download with `iers.config.auto_download=False` (but this has to be done in the astropy config, since even importing `iers` already causes problems...).

taldcroft commented on astropy/table/serialize.py:
```diff
@@ -7,6 +7,7 @@
 from .column import Column
 from .table import Table, QTable, has_info_class
 from astropy.units.quantity import QuantityInfo
+from astropy.coordinates import representation as coorep
```
After much flailing I finally figured out that this line is what is breaking things and causing all the confusing problems for me with warnings and test fails and leap seconds re-downloading every time.  It is tricky because of circular imports, so the problem only shows up in a certain import order.  You should be able to reproduce with a single `import astropy.io.ascii` without anything first.  But `import astropy.coordinates; import astropy.io.ascii` is fine.  And `import astropy.table` first gives `/Users/aldcroft/git/astropy/astropy/table/column.py:1020: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = getattr(super(), op)(other)`

This is latter warning is from the leap second stuff, and goes away if I remove this line here in `serialize.py`.

mhvk replied:
Darn, that must be because it imports `Time` - I'm sorry that I've caught all this mess, @pllim already had to move other imports inline to get things to pass.

taldcroft replied:
I suggest a quick workaround to hardcode as strings all the representations currently in astropy.

mhvk replied:
OK, will make a PR. Clearly trying to be clever comes back to haunt oneself...

mhvk replied:
See #10282
